src/core/ext/transport/binder/wire_format/wire_writer.cc

```diff
@@ -66,119 +80,301 @@ absl::Status WireWriterImpl::WriteTrailingMetadata(const Transaction& tx,
+      BinderTransportTxCode(BinderTransportTxCode::ACKNOWLEDGE_BYTES),
+      [num_bytes](WritableParcel* parcel) {
+        RETURN_IF_ERROR(parcel->WriteInt64(num_bytes));
+        return absl::OkStatus();
+      });
+  if (!result.ok()) {
+    gpr_log(GPR_ERROR, "Failed to make binder transaction %s",
+            result.ToString().c_str());
+  }
+  return result;
 }
 
 void WireWriterImpl::OnAckReceived(int64_t num_bytes) {
-  grpc_core::MutexLock lock(&mu_);
-  num_acknowledged_bytes_ = std::max(num_acknowledged_bytes_, num_bytes);
-  cv_.Signal();
+  // Ensure combiner will be run if this is not called from top-level gRPC API
+  // entrypoint.
+  grpc_core::ExecCtx exec_ctx;
+  gpr_log(GPR_INFO, "OnAckReceived %ld", num_bytes);
+  // Do not try to obtain `write_mu_` in this function. NDKBinder might invoke
+  // the callback to notify us about new incoming binder transaction when we are
+  // sending transaction. i.e. `write_mu_` might have already been acquired by
+  // this thread.
+  {
+    grpc_core::MutexLock lock(&flow_control_mu_);
+    num_acknowledged_bytes_ = std::max(num_acknowledged_bytes_, num_bytes);
+    int64_t num_outgoing_bytes = num_outgoing_bytes_;
+    if (num_acknowledged_bytes_ > num_outgoing_bytes) {
+      gpr_log(GPR_ERROR,
+              "The other end of transport acked more bytes than we ever sent, "
+              "%ld > %ld",
+              num_acknowledged_bytes_, num_outgoing_bytes);
+    }
+  }
+  TryScheduleTransaction();
+}
+
+void WireWriterImpl::TryScheduleTransaction() {
+  while (true) {
+    grpc_core::MutexLock lock(&flow_control_mu_);
+    // An estimation of number of bytes of non-acked traffic we will eventually
+    // have, assuming all tasks in combiner will be executed and we receive no
+    // new ACK from the other end of transport.
+    int64_t num_non_acked_bytes_possible =
+        (num_outgoing_bytes_ + num_non_acked_tx_in_combiner_ * kBlockSize) -
+        num_acknowledged_bytes_;
+    if (num_non_acked_bytes_possible < 0) {
+      gpr_log(GPR_ERROR,
+              "Something went wrong. `num_non_acked_bytes_possible` should be "
+              "non-negative but it is %ld",
+              num_non_acked_bytes_possible);
+    }
+    if (!pending_outgoing_tx_.empty() &&
+        (num_non_acked_bytes_possible + kBlockSize < kFlowControlWindowSize)) {
+      num_non_acked_tx_in_combiner_++;
+      combiner_->Run(GRPC_CLOSURE_CREATE(RunScheduledTx,
+                                         pending_outgoing_tx_.front(), nullptr),
+                     GRPC_ERROR_NONE);
+      pending_outgoing_tx_.pop();
+    } else {
+      if (!pending_outgoing_tx_.empty()) {
+        gpr_log(GPR_INFO,
+                "Some work cannot be scheduled yet due to slow ack from the "
+                "other end of transport. This transport might be blocked. "
+                "pending_outgoing_tx_.size() = %lu",
+                pending_outgoing_tx_.size());
+      }
+      break;
+    }
```

Comment:
This can be:

```
if (!pending_outgoing_tx_.empty()) {
  if (num_non_acked_bytes_possible + kBlockSize < kFlowControlWindowSize) {
    ...
  } else {
    ...
  }
}
```

Reply:
Done. I add `if (pending_outgoing_tx_.empty()) break;` to the beginning of the while loop